### ISSUE TITLE
centralized log aggregation

### ISSUE THREAD

**Issue by danielhartnell** [mozilla-iam/iam-infra#63]
I'm going to introduce a proof of concept this sprint.

**Issue:**

Today, container logs are retrieved with `kubectl logs`. The log content is not shipped elsewhere for analysis or data retention purposes.

**Requirements:**

I spoke with Alberto regarding this today to consider the options available to us. As far as requirements go, it would be nice if everyone in the MOC had access to this log information but I do not think it is an important requirement to meet today. It would also be nice to have some useful log analysis and dashboards presenting information queried from logs. Again, we can come back to that later.

**Proposal:**

I would like to keep things simple for now. We should be able to export the logs from the cluster and into CloudWatch Logs. The ingest rate is pretty cheap at around $0.50 / gigabyte and the cost to store data is even cheaper. Once we start to add more services to our cluster, we can get a sense for what our actual log ingest rate is and we can look at cost optimizations.

This provides us with a solution with minimal overhead. In the future, when it's appropriate, we can look at using Elasticsearch but I want to avoid the administrative overhead in the short term.

**Comment by danielhartnell** [comment]
This was pretty simple to setup. I created a new CloudWatch Logs log group called `kubernetes` and then used the following Helm chart as a reference for creating the k8s yaml:

https://github.com/helm/charts/tree/master/incubator/fluentd-cloudwatch

I filled in a few values like the region and then used the helm client template rendering to generate something that could be applied to the cluster. I also created a new role for the pod to assume with access to CloudWatch.

I had to set the `FLUENT_UID` environment variable to 0 in order to get it to function. It returns an error about being unable to open certain files if I don't do that.

Todo:

- [ ] Setup IAM role via Terraform (and restrict access - I tested with full admin access)
- [ ] Push Kubernetes YAML to this repository
- [ ] See if there are any log retention settings I should adjust in CloudWatch

**Comment by danielhartnell** [comment]
https://github.com/mozilla-iam/eks-deployment/pull/74 added the Fluentd pods.

**Comment by The-smooth-operator** [comment]
During the DinoPark workweek I deployed a new central logging stack using FluentD+Graylog+ES (AWS). 
Seems we are happy with the setup, and it already helped us finding problems in the cluster.

For the moment I wrote documentation about the stack in the respective folder where the Kubernetes resources are located. This is only missing examples on how to use it. I plan to add them in the near future, once the cluster has more applications running on it.

Still there are things to be done:
- [ ] Setup authentication via Auth0
- [ ] Add more streams (for example NGINX logs)
- [ ] Add example use-cases in the docs
- [ ] Expose the web application to outside?
- [ ] Configure different permission levels for developers, operators, etc...
